cores/nRF5/WInterrupts.c

```diff
@@ -95,25 +96,63 @@ int attachInterrupt(uint32_t pin, voidFuncPtr callback, uint32_t mode)
   }
 
-  for (int ch = 0; ch < NUMBER_OF_GPIO_TE; ch++) {
-    if (channelMap[ch] == -1 || (uint32_t)channelMap[ch] == pin) {
-      channelMap[ch] = pin;
-      callbacksInt[ch] = callback;
-      callbackDeferred[ch] = deferred;
-
-      NRF_GPIOTE->CONFIG[ch] &= ~(GPIOTE_CONFIG_PORT_PIN_Msk | GPIOTE_CONFIG_POLARITY_Msk);
-      NRF_GPIOTE->CONFIG[ch] |= ((pin << GPIOTE_CONFIG_PSEL_Pos) & GPIOTE_CONFIG_PORT_PIN_Msk) |
-                              ((polarity << GPIOTE_CONFIG_POLARITY_Pos) & GPIOTE_CONFIG_POLARITY_Msk);
-
-      NRF_GPIOTE->CONFIG[ch] |= GPIOTE_CONFIG_MODE_Event;
-
-      NRF_GPIOTE->INTENSET = (1 << ch);
-
-      return (1 << ch);
+  // All information for the configuration is known, except the prior values
+  // of the config register.  Pre-compute the mask and new bits for later use.
+  //     CONFIG[n] = (CONFIG[n] & oldRegMask) | newRegBits;
+  //
+  // Three fields are configured here: PORT/PIN, POLARITY, MODE
+  const uint32_t oldRegMask = ~(GPIOTE_CONFIG_PORT_PIN_Msk | GPIOTE_CONFIG_POLARITY_Msk | GPIOTE_CONFIG_MODE_Msk);
+  const uint32_t newRegBits =
+    ((pin                      << GPIOTE_CONFIG_PSEL_Pos    ) & GPIOTE_CONFIG_PORT_PIN_Msk) |
+    ((polarity                 << GPIOTE_CONFIG_POLARITY_Pos) & GPIOTE_CONFIG_POLARITY_Msk) |
+    ((GPIOTE_CONFIG_MODE_Event << GPIOTE_CONFIG_MODE_Pos    ) & GPIOTE_CONFIG_MODE_Msk    ) ;
+
+  int ch = -1;
+  int newChannel = 0;
+
+  // Find channel where pin is already assigned, if any
+  for (int i = 0; i < NUMBER_OF_GPIO_TE; i++) {
+    if ((uint32_t)channelMap[i] != pin) continue;
+    ch = i;
+    break;
+  }
+  // else, find one not already mapped and also not in use by others
+  if (ch == -1) {
+    for (int i = 0; i < NUMBER_OF_GPIO_TE; i++) {
+      if (channelMap[i] != -1) continue;
+      if (nrf_gpiote_te_is_enabled(NRF_GPIOTE, i)) continue;
+      
+      ch = i;
+      newChannel = 1;
+      break;
     }
   }
+  // if no channel found, exit
+  if (ch == -1) {
+    return 0; // no channel available
+  }
 
-  return 0;
+  uint32_t tmp = NRF_GPIOTE->CONFIG[ch];
+  tmp &= oldRegMask;
+  tmp |= newRegBits;                 // for existing channel, effectively updates only the polarity
+  channelMap[ch]         = pin;      // harmless for existing channel
+  callbacksInt[ch]       = callback; // caller might be updating this for existing channel
+  callbackDeferred[ch]   = deferred; // caller might be updating this for existing channel
+  NRF_GPIOTE->CONFIG[ch] = tmp;
+
+  // For a new channel, additionally ensure no old events existed, and enable the interrupt
+  if (newChannel) {
+    // Why the memory barrier and NOPs? 
+    // See page 4 at https://web.archive.org/web/20190823222657/http://infocenter.arm.com/help/topic/com.arm.doc.dai0321a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf
+    // and recall the AHB runs at 16MHz, so may take 4 cycles before peripheral register is actually updated.
+    __DSB(); __NOP();__NOP();__NOP();__NOP();
```

Comment:
please remove these _DSB(), the order of these 2 are not important. Plus I don't see the reason why gcc would run it differently.

Reply:
see response to [next comment](https://github.com/adafruit/Adafruit_nRF52_Arduino/pull/485#discussion_r416054396)